### ISSUE TITLE
getSeenRequests never completing due to rule.requests never resolving promise.

### ISSUE THREAD

**Issue by hanvyj** [httptoolkit/mockttp#57]
I've got a mocked endpoint:
```
    const mockedEndpoint = mockServer.post("/id/endpoint").thenJson(200 , { total: 3 });
```
I can see with the logging it's being added okay:

>Handling request for /id/endpoint
>Request matched rule: Match requests making POSTs for /id/endpoint, and then respond with status 200, headers {"Content-Type":"application/json","Content-Length":"40"} and body "{"total":3}".

The endpoint is called in a browser test, and it does respond with the correct reply, within 6ms, not stuck pending.

Then I later check it's been called with the correct info:

```
log.debug("Got mocked endpoint", mockedEndpoint);
const requests = await mockedEndpoint.getSeenRequests();
```
The problem is this times out. The promise never completes. I added some logging to ensure I've got the right endpoint:

> Got mocked endpoint Mocked endpoint: Match requests making POSTs for /id/endpoint, and then respond with status 200, headers {"Content-Type":"application/json","Content-Length":"40"} 
and body "{"total":3}".

Then I added some logging to the library itself to check the requests promises:
```
    getSeenRequests() {
        // Wait for all completed running requests to have all their details available
        console.log(this.rule.requests);
        return Promise.all(this.rule.requests);
    }
```

Which reports the single expected request that's stuck: `[ Promise { <pending> } ]`

I've no idea why this particular endpoint is hanging, it works with all the others (more complex ones with body and query matchers). The response does come through in the browser tests where I'm calling it. It responds with the correct reply, within 6ms. It's just the call to getSeenRequests.

Have you seen this before? Is there anything that would case it to get stuck unresolved?

**Comment by pimterry** [comment]
Hi @hanvyj. I'm not sure what could cause this actually! If that's happening I'd definitely like to fix it though. Is it possible to share a quick example so I can reproduce this?

You might be able to see more in the browser dev tools. Are there any messages printed there? That should be an internal request that's sent to query the latest seen requests when that `getSeenRequests` is called too - is that sent, and does it get a response?

**Comment by hanvyj** [comment]
I'm not sure I could isolate it to a reproducible example. It was working absolutely fine before a small change. I've added a special edge case that isolates the particular request (that checks if there's a request body or query and adds those) and simply runs:

```
mockServer.post(request.path).thenJson(response.status, response.body)
```

But when I isolate this to a test:

```
        it("Edge case getSeenRequests test", async () => {
            const mockedEndpoint = await server.post("/id/endpoint").thenJson(200 , { total: 3 });

            const response = await fetch(`${server.url}/id/endpoint`, {
                method: 'POST',
                body: '{"query":"","reason":"Investigated","comment":""}'
            })

            const requests = await mockedEndpoint.getSeenRequests();

            console.log(response.status);
            console.log(requests);
        });
```

It's absolutely fine!

There's not much reported in the browser, no comments, the request is a 200 and returns the correct data response.

![image](https://user-images.githubusercontent.com/20140222/131473930-599718f3-c92c-4432-a52f-6d14d9de0403.png)

Can you specify where the internal request is sent for `getSeenRequests`? I placed some logging in the `ServerMockedEndpoint` code:

```
export class ServerMockedEndpoint implements MockedEndpoint {

    constructor (private rule: RequestRule | WebSocketRule) {
        this.getSeenRequests.bind(this);
    };
    ...

    getSeenRequests(): Promise<CompletedRequest[]> {
        // Wait for all completed running requests to have all their details available
        console.log(this.rule.requests); // I added logging here to get "[ Promise { <pending> } ]"
        return Promise.all<CompletedRequest>(this.rule.requests);
    }
```

I tried adding logging to `MockedEndpointClient` which uses the `endpointDataGetter` and appears to send a query:

```
private getEndpointDataGetter = (ruleId: string) => async (): Promise<MockedEndpointData | null> => {
        let result = await this.queryMockServer<{
            mockedEndpoint: MockedEndpointData | null
        }>(
            `query GetEndpointData($id: ID!) {
                mockedEndpoint(id: $id) {
                    seenRequests {
                        protocol,
                        method,
                        url,
                        path,
                        hostname,
                        headers,
                        body,
                        ${this.optionalField('Request', 'timingEvents')}
                        ${this.optionalField('Request', 'httpVersion')}
                    }
                    ${this.optionalField('MockedEndpoint', 'isPending')}
                }
            }`, {
                id: ruleId
            }
        );

        const mockedEndpoint = result.mockedEndpoint;

        if (!mockedEndpoint) return null;

        mockedEndpoint.seenRequests.forEach((request: any) => {
            request.body = buildBodyReader(Buffer.from(request.body, 'base64'), request.headers);
        });

        return mockedEndpoint;
    }
```
But when I added logging I couldn't see it, I presume this is happening in the other process that's actually processing the requests?

**Comment by hanvyj** [comment]
Interestingly, using `node-fetch` to send an identical body after setting up the endpoint:

```
    mockEndpoint = await mockServer.post(request.path).thenJson(response.status, response.body);

    await fetch(`${mockServer.url}/${request.path}`, {
      method: "POST",
      body: '{"query":"","reason":"Investigated","comment":""}',
    });

    const requests = await mockEndpoint.getSeenRequests();

    console.log(requests);
```

That works fine, and later I can see that request, the browser requests pend indefinitely. So the first request is resolved, the second never does:

```
[
  Promise {
    {
      id: 'a64e04cc-e45c-4230-b3f7-982999f3ed04',
      matchedRuleId: '00247f6c-447a-45f0-bd84-1e307e37db89',
      protocol: 'http',
      httpVersion: '1.1',
      method: 'POST',
      url: 'http://localhost:3005/aef088c4-54ff-4942-a4a1-99aee676eb6e/bulkArchive',
      path: '/aef088c4-54ff-4942-a4a1-99aee676eb6e/bulkArchive',
      headers: [Object],
      tags: [],
      timingEvents: [Object],
      body: [Object]
    }
  },
  Promise { <pending> }
]
```
So it's specific to the browser somehow?

**Comment by pimterry** [comment]
> But when I isolate this to a test ... It's absolutely fine!

Can you try to trace this further? If you run your normal test that fails by itself, does that fail?

How are you setting up the standalone & mock server? If you can share more of the code that does `server.start()`, `server.stop()` and so on that might be useful.

Some possible causes I can imagine:

* The mock server is has been stopped before `getSeenRequests()` is called.
* There's multiple rules active that are somehow conflicting.
* The standalone server is getting restarted or reset unexpectedly.

Do any of those sound possible?

> I tried adding logging to MockedEndpointClient which uses the endpointDataGetter and appears to send a query ... But when I added logging I couldn't see it.

Huh, that's very suspicious. That `endpointDataGetter` definitely should be running in the browser.

When running tests in the browser, this is how it works:

* Registering the rule returns a [MockedEndpointClient](https://github.com/httptoolkit/mockttp/blob/master/src/client/mocked-endpoint-client.ts)
* When you call `.getSeenRequests()` on that, it uses the endpointDataGetter to query the overall endpoint state.
* That getter sends a GraphQL request to the standalone server for the data.
* When that request is received, the server [calls](https://github.com/httptoolkit/mockttp/blob/dc771f23274cd66102d201fcec904107ed6789fa/src/standalone/standalone-model.ts#L210-L219) the `getSeenRequests()` method on the corresponding server-side ServerMockedEndpoint.
* The result of that gets returned as GraphQL response, and then passed back to the MockedEndpointClient, which gives you the results in the browser.

Since the ServerMockedEndpoint is being hit, the request must be being sent and handled on the server, but for some reason the response isn't making it back.

In node, none of that happens: you get a ServerMockedEndpoint as soon as you register the rule, and you're just directly calling it, so the fact that that works means it's definitely this bit of communication that's broken.

There must be some GraphQL requests in your network console in the browser happening here (because otherwise the server method wouldn't be called at all) so they must either not be completing at all or be completing with an error that isn't properly handled.

They'll be requests to `http://localhost:45456` (by default, unless you've configured your own standalone server URL). Can you check the console for those? There'll be quite a few, because there's a similar call made to register each rule as well.

**Comment by hanvyj** [comment]
> That `endpointDataGetter` definitely should be running in the browser.

I'm using mockttp for component tests using selenium. I've got a node test running, setting up the expected endpoints with mockttp, then controlling the browser through selenium. So the browser isn't running mockttp code. So mockttp is running in node, is that why I never see a `MockttpClient` and the graphql query?

At the start of the tests, running from node, I create the mock server, it's reset between each test. The endpoints are stored in a map for retrieval to call `getSeenRequests` and check endpoints were called by the selenium controlled browser:

```
/** Mock server instance */
const mockServer = getLocal({ cors: true, debug: true });

/** Mock endpoints. */
const mockedEndpointMap: Map<string, MockedEndpoint> = new Map<string, MockedEndpoint>();

/**
 * Start the mock server.
 */
export async function startMock(): Promise<void> {
  log.debug("Starting mock server");
  if (ConfigStore.getConfig().debugApi) {
    mockServer.enableDebug();
  }
  await mockServer.start(ConfigStore.getConfig().mockServer.port);
}

/**
 * Stop the mock server.
 */
export async function stopMock(): Promise<void> {
  await mockServer.stop();
}

export function clearMock() {
  mockedEndpointMap.clear();
  mockServer.reset();
}
```
I added logging to the stop and clear, but it doesn't appear to be being called. Could it stop due to an exception or something internally?

I followed the creation of the promise, in the `handle` method returns `waitForCompletedRequest(req)`, which calls `waitForBody` in `request-utils`, adding comments around this await:

```
const waitForBody = async (body: OngoingBody, headers: Headers): Promise<CompletedBody> => {
    console.log("Getting body as buffer");
    const bufferBody = yield body.asBuffer();
    console.log("Got body as buffer");
    return buildBodyReader(bufferBody, headers);
};
```
For that particular request I don't see "Got body as buffer". 


Not sure what you mean when you say:

>They'll be requests to http://localhost:45456 (by default, unless you've configured your own standalone server URL). Can you check the console for those? There'll be quite a few, because there's a similar call made to register each rule as well.

Does it being in node change where you want me to look? I'm seeing the calls to register rules, and match requests and it all looks normal.

**Comment by pimterry** [comment]
Ah, ok, that changes it completely. Disregard most of the above then, I thought you were controlling Mockttp from a browser :smile:.

`waitForBody` not resolving makes sense, I can see how that would cause this! That means that somewhere there is an open request which has sent the headers etc, but hasn't yet finished sending the request body. That could be a bug, or it could be a request that's intentionally continuously streaming data.

If you check inside the browser, are there any pending requests like that? You could also add logging at the start & end of waitForCompletedRequest to log the request URL, and check exactly which request hasn't sent its body yet.

This should only happen for requests that match this rule and then persist in that incomplete state. It's just about plausible that there could be a bug where this waits for all outstanding requests though... If you see this being caused by a pending request that shouldn't match this rule but which is still interfering then I can investigate further.

I do think this is probably the right behaviour for `getSeenRequests` (giving you the requests only once their content is fully resolved, not returning partial requests) but if this is happening because of some intentional behaviour and there's a real use case where different behaviour would be helpful, I'm open to extending this to support that somehow.

**Comment by hanvyj** [comment]
I ran a `mockServer.getPendingEndpoints()` and it didn't list anything (though I think that only tracks endpoints that haven't been called at least once?)

There are no calls from the browser that aren't 503s for unmatched (I've not bothered mocking everything for this particular test), or 200 for a response. I'm not sure what could cause them to keep streaming. None are pending:

![image](https://user-images.githubusercontent.com/20140222/131502504-2e3a9181-e25c-4a41-b2ff-7865f77befda.png)

I've added some logging around the streaming and promises, excuse the random characters, it just helps match to where I've shoved things in the js:

```
asBuffer() {
            console.log("DDDDD asBuffer ", bufferPromise)
            if (!bufferPromise) {
                console.log("DDDDD Streaming to buffer")
                bufferPromise = buffer_utils_1.streamToBuffer(bodyStream, maxSize);
                console.log("DDDDD Promise", bufferPromise)
                bufferPromise
                    .then((buffer) => {
                        console.log("DDDDD COMPLETED PROMISE", buffer.toString());
                        completedBuffer = buffer;
                    })
                    .catch(() => { }); // If we get no body, completedBuffer stays null
            }
            console.log("DDDDD Returning ", bufferPromise)
            return bufferPromise;
        },
```
And in buffer-utils, `streamToBuffer`:
```
input.on('data', (d) => {
            console.log("EEEEE streamToBuffer data", d);
            currentSize += d.length;
            // If we go over maxSize, drop the whole stream, so the buffer
            // resolves empty. MaxSize should be large, so this is rare,
            // and only happens as an alternative to crashing the process.
            if (currentSize > maxSize) {
                console.log("EEEEE streamToBuffer max size");
                chunks = []; // Drop all the data so far
                return; // Don't save any more data
            }
            chunks.push(d);
        });
        input.once('end', () => {
            console.log("EEEEE Ended");
            resolve(Buffer.concat(chunks));
        });
        input.once('aborted', () => {
            console.log("EEEEE Aborted");
            bufferPromise.failedWith = new Error('Aborted');
            reject(bufferPromise.failedWith);
        });
        input.on('error', (e) => {
            console.log("EEEEE Error");
            bufferPromise.failedWith = bufferPromise.failedWith || e;
            reject(e);
        });
```

I get the response:

```
Handling request for /289842bc-e4fa-44da-8d60-8eabea445f27/bulkArchive
Request matched rule: Match requests making POSTs for /289842bc-e4fa-44da-8d60-8eabea445f27/bulkArchive, and then respond with status 200, headers ....
AAAAAA HANDLING REQUEST  d9839354-be11-4868-ac9e-b254acd26406
AAAAAA PUSHING COMPLETED REQUEST Promise { <pending> }
AAAAAA ADDING REQUEST COUNT
AAAAAA COMPLETED PROMISE Promise { <pending> }
AAAAAA WAITING FOR COMPLETION  d9839354-be11-4868-ac9e-b254acd26406
BBBBB Waiting for body
CCCCC Getting body as buffer
DDDDD asBuffer  null
DDDDD Streaming to buffer
DDDDD Promise Promise { <pending>, currentChunks: [] }
DDDDD Returning  Promise { <pending>, currentChunks: [] }
BBBBB Waiting for body
CCCCC Getting body as buffer
DDDDD asBuffer  Promise { <pending>, currentChunks: [] }
DDDDD Returning  Promise { <pending>, currentChunks: [] }
```

Interestingly, I also get a similar call to another endpoint just before it:
```
Handling request for /289842bc-e4fa-44da-8d60-8eabea445f27/cluster
Request matched rule: Match requests making POSTs for /289842bc-e4fa-44da-8d60-8eabea445f27/cluster, and then respond with status 200, headers ....
AAAAAA HANDLING REQUEST  6fa7e561-262c-4059-8703-7a7c458c2582
AAAAAA PUSHING COMPLETED REQUEST Promise { <pending> }
AAAAAA ADDING REQUEST COUNT
AAAAAA COMPLETED PROMISE Promise { <pending> }
AAAAAA WAITING FOR COMPLETION  6fa7e561-262c-4059-8703-7a7c458c2582
BBBBB Waiting for body
CCCCC Getting body as buffer
DDDDD asBuffer  null
DDDDD Streaming to buffer
DDDDD Promise Promise { <pending>, currentChunks: [] }
DDDDD Returning  Promise { <pending>, currentChunks: [] }
BBBBB Waiting for body
CCCCC Getting body as buffer
DDDDD asBuffer  Promise { <pending>, currentChunks: [] }
DDDDD Returning  Promise { <pending>, currentChunks: [] }
```
It seems both the successfully matched requests never see a `nput.on('data'` or `input.once('end'`.

For the unmatched responses I can see this:

```
Unmatched request received: GET request to http://localhost:3005/289842bc-e4fa-44da-8d60-8eabea445f27/searchTermsAggregation/CLUSTER/threatTypes.name.keyword with headers:...
DDDDD asBuffer  Promise { <Buffer >, currentChunks: [] }
DDDDD Returning  Promise { <Buffer >, currentChunks: [] }
BBBBB Waiting for body
CCCCC Getting body as buffer
DDDDD asBuffer  Promise { <Buffer >, currentChunks: [] }
DDDDD Returning  Promise { <Buffer >, currentChunks: [] }
CCCCC Got body as buffer
BBBBB Got BODY
BBBBB buildInitiatedRequest
BBBBB Returning
CCCCC Getting body as buffer
DDDDD asBuffer  null
DDDDD Streaming to buffer
DDDDD Promise Promise { <pending>, currentChunks: [] }
DDDDD Returning  Promise { <pending>, currentChunks: [] }
EEEEE streamToBuffer data <Buffer 4e 6f 20 72 75 6c 65 73 20 77 65 72 65 20 66 6f 75 6e 64 20 6d 61 74 63 68 69 6e 67 20 74 68 69 73 20 72 65 71 75 65 73 74 2e 0a 54 68 69 73 20 72 65 ... 6752 more bytes>
EEEEE Ended
DDDDD COMPLETED PROMISE No rules were found matching this request. ...(lots more)
```

**Comment by pimterry** [comment]
Is there any way you can share more info so I can reproduce this myself? It's very difficult to debug like this! Happy to delve into the whole project, if the source is available, I'd really like to know the cause of this.

I can think of a few possibilities that might do this:

1. Maybe there's some odd browser behaviour, where because Mockttp replies immediately (thenJson will reply before the request body is received) the browser never sends the request body at all, but also somehow never errors or aborts? This should be valid (servers have no obligation to wait for the request body) but it's just about possible that could do weird things.
2. Maybe something inside Mockttp is incorrectly reading the whole stream before it gets to waitForCompletedRequest, so it's already finished?
3. Maybe there really is a pending request, but its coming from the browser itself, not inside the page, and that's causing issues?

A few things that might differentiate this:

* Add a `withBodyIncluding(substring)` matcher to your rule - this will wait for the entire body, and check that it matches, before the response is sent. If that resolves the issue then option 1 looks plausible - sending the request early is causing problems.
* Check the values of `input.readable`, `input.readableEnded` and `input.readableAborted` inside `streamToBuffer`. If the readable was already ended or aborted before the first call to that method then there's definitely a bug there and option 2 looks likely.
* Try shutting down Selenium & the browser while the test is running, and see if that closes the stream so that the promise resolved. If that by itself reliably resolves the promise then there really is a pending request somewhere and it's definitely option 3.

**Comment by hanvyj** [comment]
> Add a withBodyIncluding(substring) matcher to your rule - this will wait for the entire body, and check that it matches, before the response is sent. If that resolves the issue then option 1 looks plausible - sending the request early is causing problems

That did it! `.withBodyIncluding("")` solved it.

The 'cause' of the problem for me was changing `.thenReply(response.status, JSON.stringify(response.body))` to `.thenJson(response.status, response.body)`. Changing it back to `thenReply` also fixes it, does that also wait for the entire body before sending the response?

I put it back to `thenJson` and closing the browser did not resolve the promise!

Yes, it's very frustrating trying to debug like this. Sadly the project isn't open source and I can't share the code (it's also huge and difficult to pair down to an example). We're hoping to maybe open source our cucumber/selenium based component testing framework at some point, I might try push to do that sooner so I can actually share it. I've failed to reproduce it by using the exact same `mockServer.post(request.path).thenJson(response.status, response.body)` and then calling it in chrome (without selenium).

**Comment by pimterry** [comment]
> The 'cause' of the problem for me was changing .thenReply(response.status, JSON.stringify(response.body)) to .thenJson(response.status, response.body). 

That is _very_ surprising. Are you totally sure?

You can compare the implementations here:

https://github.com/httptoolkit/mockttp/blob/2dffcd9b036c5bba21c54b711aba84881831f510/src/rules/requests/request-rule-builder.ts#L111-L117

https://github.com/httptoolkit/mockttp/blob/2dffcd9b036c5bba21c54b711aba84881831f510/src/rules/requests/request-rule-builder.ts#L136-L149

After the initial parameter handling for thenReply (to allow an optional status message), there's not a huge difference, it's just that:

* `thenJson` does the stringify automatically.
* `thenJson` adds a content-length & content-type to the response headers.

I suppose it's possible that the content length causes issues here, since that could change exactly how the response & connection is handled afterwards... By default I think Node will use chunked encoding, rather than an explicit length.

Assuming you can reliably reproduce it, can you check if commenting out [that extra content-length header](https://github.com/httptoolkit/mockttp/blob/2dffcd9b036c5bba21c54b711aba84881831f510/src/rules/requests/request-rule-builder.ts#L140) fixes it for you?

I'll do some more investigation myself and see if I can find a way to reproduce and fix this more thoroughly.

**Comment by hanvyj** [comment]
Yes, I was equally confused. I'd overlooked the change to `thenJson` for ages because I didn't think that change could possibly have caused any issues.

I also looked at the difference, and I've found 

```
.thenReply(response.status, JSON.stringify(response.body), { 'Content-Type': 'application/json' });
```
Didn't work. So I don't think it's the content-length. I did try it and it also doesn't work:

```
.thenReply(response.status, JSON.stringify(response.body), { 'Content-Length': Buffer.from(JSON.stringify(response.body), 'utf8').byteLength });
```

Finally, I thought I'd just try any header:
```
.thenReply(response.status, JSON.stringify(response.body), { 'test-header': "value" })
```
Also fails!

Following it up, if I comment out the line `dropDefaultHeaders(response)`:

```
    async handle(_request: OngoingRequest, response: OngoingResponse) {
        if (this.headers) {
            dropDefaultHeaders(response);
            setHeaders(response, this.headers);
        }
        response.writeHead(this.status, this.statusMessage);

        if (isSerializedBuffer(this.data)) {
            this.data = Buffer.from(<any> this.data);
        }

        response.end(this.data || "");
    }
```
It works. Following it further, if I prevent the "transfer-encoding" header being removed, it seems to work fine:

```
// If the user explicitly specifies headers, we tell Node not to handle them,
// so the user-defined headers are the full set.
export function dropDefaultHeaders(response: OngoingResponse) {
    // Drop the default headers, so only the headers we explicitly configure are included
    [
        'connection',
        'content-length',
        // 'transfer-encoding',
        'date'
    ].forEach((defaultHeader) =>
        response.removeHeader(defaultHeader)
    );
}
```

**Comment by pimterry** [comment]
Aha, ok, yes I'd forgotten that setting headers means you're setting _all_ headers.

That fits though: this all works when the response is sent as a chunked response (with a transfer encoding from the default headers) but for some reason when it's sent immediately with an explicit content-length then something happens that means Mockttp never captures the request body.

That kind of pins down the trigger for this, but I still don't understand why the last step would happen.

Any chance you could capture a successful and unsuccessful set of requests & responses with [Wireshark](https://www.wireshark.org/), and then share the resulting PCAPs? I think this is going to come down to low-level connection handling somewhere, so being able to explore the raw packets will probably have some clues. If you want to make sure that just relevant traffic is included you can filter to only the relevant packets (e.g. `http && ip.addr == 127.0.0.1`) and then export via File -> Export Specified Packets.

Thanks for all your hard work tracing this down by the way! I really appreciate it. Just FYI, [HTTP Toolkit Pro](https://httptoolkit.tech/pricing/) is totally free for all contributors to everything in this GH org, and this definitely counts (HTTP Toolkit is built on top of Mockttp). Just let me know your email if you want a free account some time.

**Comment by hanvyj** [comment]
I'm not familiar with many of these headers, but I randomly experimented with which is causing the issue. If I remove just the 'connection' header, it displays the issue I've seen. If I remove just the transfer-encoding header, the requests don't even respond (they're pending I think) until the mockttp server is closed.

Removing neither the 'connection' and 'transfer-encoding' and it functions:
```
    [
        // 'connection',
        'content-length',
        // 'transfer-encoding',
        'date'
    ].forEach((defaultHeader) =>
        response.removeHeader(defaultHeader)
    );
```

Still can't reproduce by setting those headers though. I'll take a look at wireshark.

**Comment by hanvyj** [comment]
Here's two runs of the same test. One with 'connection',  'transfer-encoding' headers 'removed' (`failing-request.pcapng`) that displays the behaviour. One with those two header removals commented out (`working-request.pcapng`) which doesn't hang when calling `getSeenRequests()`.

[wireshark.zip](https://github.com/httptoolkit/mockttp/files/7091853/wireshark.zip)

**Comment by pimterry** [comment]
I've looked into those traces, and I can see the request differences, but I can't see anything actually wrong there, certainly nothing that would cause this issue. Both content-length & chunked encoding should be equally valid, and there's no reason sending a keep alive response would block receiving the request body!

I will continue looking into this though. I'd really like to get this sorted, I'll let you if I find any more clues or if I manage to reproduce this myself.

**Comment by pimterry** [comment]
Hey @hanvyj - I've just released 2.3.0, which includes your changes from #56, but also a few relevant fixes around `getSeenRequests` for a various cases where that could error and timeout unexpectedly.

These are good general fixes regardless, but I'm also hopeful that they might relate to your mystery issue here. When you get a minute at some point, can you update and see if the issue where `thenJson` stopped `getSeenRequests` from resolving is now fixed?

**Comment by pimterry** [comment]
I'm going to assume this is now fixed and close this, but feel free to share more info if not.
